File: cs/Problems/Day9.cs
namespace aoc24.Problems;

public sealed class Day9 : IProblem<long>
{
    public long Solve(string input) => CalculateDiskChecksumOptimized(input);

    const int EMPTY_SPACE = -1;

    public static long CalculateDiskChecksumOptimized(ReadOnlySpan<char> input)
    {
        // Expand file volume... O.O
        Span<int> expanded = stackalloc int[input.Length * 5];
        int size = 0;

        for (int i = 0; i < input.Length; i++)
        {
            int positions = input[i] - '0';
            int fileId = i % 2 == 0 ? i / 2 : EMPTY_SPACE;

            for (int s = 0; s < positions; s++, size++)
                expanded[size] = fileId;
        }

        // Calculate checksum...
        // In theory, we should reorganize free space first, but I figured I could do this in one go.
        long checksum = 0;

        for (int head = 0, tail = size; head < tail; head++)
        {
            int value = expanded[head];

            if (value == EMPTY_SPACE)
            {
                while (--tail > head)
                {
                    value = expanded[tail];

                    if (value != EMPTY_SPACE)
                        break;
                }
            }

            checksum += value * head;
        }

        return checksum;
    }

    public static long CalculateDiskChecksum(string input)
    {
        var expanded = new List<int>();

        for (int i = 0; i < input.Length; i += 2)
        {
            int files = input[i] - '0';
            int space = i < input.Length - 1 ? input[i + 1] - '0' : 0;
            int fileId = i / 2;

            expanded.AddRange(Enumerable.Repeat(fileId, files));
            expanded.AddRange(Enumerable.Repeat(-1, space));
        }

        long checksum = 0;

        for (int head = 0, tail = expanded.Count; head < tail; head++)
        {
            int value = expanded[head];

            if (value == EMPTY_SPACE)
            {
                while (--tail > head)
                {
                    value = expanded[tail];

                    if (value != EMPTY_SPACE)
                        break;
                }
            }

            checksum += value * head;
        }

        return checksum;
    }
}

File: go/problems/day9.go
package problems

type Day9 struct{}

var emptySpace int = -1

func (d Day9) Solve(input string) int64 {
	universe := make([]int, 0, len(input))

	for i := 0; i < len(input); i += 2 {
		fileId := i / 2
		files := int(input[i] - '0')
		spaces := 0
		if i < len(input)-1 {
			spaces = int(input[i+1] - '0')
		}

		ul := len(universe)
		universe = append(universe, make([]int, files+spaces)...)

		for i := range files {
			universe[ul+i] = fileId
		}
		for i := range spaces {
			universe[ul+files+i] = emptySpace
		}
	}

	var r int64 = 0

	for head, tail := 0, len(universe); head < tail; head++ {
		val := universe[head]

		if val == emptySpace {
			tail--
			for tail > head {
				val = universe[tail]
				if val != emptySpace {
					break
				}
				tail--
			}
		}

		r += int64(val * head)
	}

	return r
}
